agent/engine/common_integ_testutil.go

```diff
@@ -259,6 +262,8 @@ func SetupIntegTestTaskEngine(cfg *config.Config, state dockerstate.TaskEngineSt
 	taskEngine := NewDockerTaskEngine(cfg, dockerClient, credentialsManager,
 		eventstream.NewEventStream("ENGINEINTEGTEST", context.Background()), imageManager, &hostResourceManager, state, metadataManager,
 		nil, execcmd.NewManager(), engineserviceconnect.NewManager(), daemonManagers)
+	taskEngine.taskSteadyStatePollInterval = 1 * time.Second
+	taskEngine.taskSteadyStatePollIntervalJitter = 500 * time.Millisecond
```

Comment:
When a task enters the steady state, it will wait the full duration of `steadyStatePollInterval + steadyStatePollIntervalJitter` if no ACS/Docker event or resource state change happens first.
https://github.com/aws/amazon-ecs-agent/blob/e37d2c0bb910177c9136bc1c20a5370088dd388b/agent/engine/task_manager.go#L332-L335
When the task stopper [changes the tasks's desired state to `STOPPED`](https://github.com/aws/amazon-ecs-agent/blob/e37d2c0bb910177c9136bc1c20a5370088dd388b/agent/acs/session/task_stop_verification_ack_responder.go#L44-L45), no ACS/Docker event or resource state change is emitted, so the task ends up waiting the full poll interval duration before progressing to the `STOPPED` state. This can be ~5 minutes if the default `steadyStatePollInterval` value is used, so I've lowered the values in tests to speed them up.

Reply:
Thanks for the additional info. If this behavior exists in Agent then that sounds concerning to me. Can we please check if indeed no ACS events are emitted when this workflow is triggered via an ACS message?

Reply:
It appears that an ACS event is not pushed onto the task's `acsMessages` channel when Agent receives a `task stop ACK` message.
https://github.com/aws/amazon-ecs-agent/blob/e37d2c0bb910177c9136bc1c20a5370088dd388b/agent/engine/task_manager.go#L381-L387
The only place where we write to the `acsMessages` channel is in `emitACSTransition`.
https://github.com/aws/amazon-ecs-agent/blob/e37d2c0bb910177c9136bc1c20a5370088dd388b/agent/engine/task_manager.go#L785-L793
`emitACSTransition` is only called by [`DockerTaskEngine.updateTaskUnsafe`](https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/docker_task_engine.go#L2687), and `DockerTaskEngine.updateTaskUnsafe` is only called by [`Engine.AddTask`](https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/docker_task_engine.go#L1210).

Reply:
I believe we can add `ts.taskEngine.AddTask(task)` after these 2 lines to send an ACS event for `test stop ACK`s.
https://github.com/aws/amazon-ecs-agent/blob/e37d2c0bb910177c9136bc1c20a5370088dd388b/agent/acs/session/task_stop_verification_ack_responder.go#L44-L45

With these changes, the integration tests don't hang when the steady state poll interval is 5 minutes.
https://github.com/tinnywang/amazon-ecs-agent/compare/task_stop_integ_tests...tinnywang:amazon-ecs-agent:stop_task_faster?expand=1

Reply:
Thanks for figuring this out! Yeah I think we should add back the `AddTask` call.